### PR TITLE
Better angular 1.4.7 types

### PR DESCRIPTION
**Pull request by villesau**
Plenty of enhancements for angular 1.x libdefs. These are the types we currently use at Smartly for angular. So not perfect, but better. Some utility function types missing etc.
Some of the commits have link to appropriate angular documentation to make reviewing slightly easier.

ping @LoganBarnett

### CODE REVIEW

File: definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js [hunk 1]
```diff
@@ -7,9 +7,16 @@
...
+  declare type ScopeBindings = '<' | '=' | '&' | '<?' | '=?' | '&?' | '=ngModel';
+  declare type Scope = {[key: string]: ScopeBindings};
+  declare type ControllerFunction = (...a: Array<*>) => void;
+
   // I'm not sure how represent this properly: Angular DI declarations are a
   // list of strings with a single function at the end. The function can vary,
   // so it is a type param.
+  //
+  // NOTE: if you use compile step to mangle array, replace below with
+  // declare type $npm$angular$DependencyInjection<T> = T
```
Comment (villesau) [inline definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js]: This is something i would like to be able to handle in better way than just writing a comment here. We don't use array dependency injection syntax with angular, and i think it's pretty common approach.

File: definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js [hunk 2]
```diff
@@ -22,12 +29,12 @@ declare module angular {
...
 
-  declare function AngularLinkFunction(
-    scope: mixed,
+  declare type AngularLinkFunction = (
+    scope: $Scope<*>,
     element: JqliteElement,
     attrs: mixed,
     controller: mixed
-  ): void
+  ) => void
```
Comment (villesau) [inline definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js]: any idea why flow couldn't find `$Scope<*>` when AngularLinkFunction was declared as `declare function`? And whats the actual difference with `declare function fn():void` and `declare type fn = () => void`?

Comment (LoganBarnett) [reply definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js]: I've found it's easier to declare overrides with `declare type` and just intersection them with `&`. Could that be related to what you're looking at?

Comment (villesau) [reply definitions/npm/angular_v1.4.7/flow_v0.47.x-/angular_v1.4.7.js]: Seems that with `declare function` ordering matters, but not with `declare type fn = () => {}`. Do you know if there are any other differences? One what i'm really curious about is that is it somehow possible to type function `this`. Naturally that's not possible with arrow function, but with regular function libdefs that would make much sense.

**Comment by GAntoine** [comment]
@villesau Ping me when these are ready to merge.

**Comment by villesau** [comment]
@GAntoine added type for angular $timeout. Good to go from my perspective.

**Pull request merged**